NLL: E0594 removes first char of variable name

When I build this code on nightly...

```rust
#![feature(nll)]

fn main() {
    let foo = &16;
    *foo = 32;
    let bar = foo;
    *bar = 64;
}
```

...I get these errors:

```none
error[E0594]: cannot assign to data in a `&` reference
 --> src/main.rs:5:5
  |
4 |     let foo = &16;
  |               --- help: consider changing this to be a mutable reference: `&mut 16`
5 |     *foo = 32;
  |     ^^^^^^^^^

error[E0594]: cannot assign to data in a `&` reference
 --> src/main.rs:7:5
  |
6 |     let bar = foo;
  |               --- help: consider changing this to be a mutable reference: `&mut oo`
7 |     *bar = 64;
  |     ^^^^^^^^^
```

The first error is fine, but the second one is missing the first character of `foo`. It looks like rustc assumes the first character is an `&` and removes it without checking.

## Meta

`rustc --version --verbose`:
```none
rustc 1.28.0-nightly (1d4dbf488 2018-06-11)
binary: rustc
commit-hash: 1d4dbf488a4364413e7b9611866b7a5c75ce566f
commit-date: 2018-06-11
host: x86_64-unknown-linux-gnu
release: 1.28.0-nightly
LLVM version: 6.0
```

--- a/src/librustc_mir/borrow_check/mod.rs
+++ b/src/librustc_mir/borrow_check/mod.rs
@@ -1905,8 +1905,10 @@ impl<'cx, 'gcx, 'tcx> MirBorrowckCtxt<'cx, 'gcx, 'tcx> {
                     // highlighted text will always be `&<expr>` and
                     // thus can transform to `&mut` by slicing off
                     // first ASCII character and prepending "&mut ".
-                    let borrowed_expr = src[1..].to_string();
-                    return (assignment_rhs_span, format!("&mut {}", borrowed_expr));
+                    if src.starts_with('&') {
+                        let borrowed_expr = src[1..].to_string();
+                        return (assignment_rhs_span, format!("&mut {}", borrowed_expr));
+                    }
                 }
             }
 

--- a/src/test/ui/suggestions/issue-51515.rs
+++ b/src/test/ui/suggestions/issue-51515.rs
@@ -12,7 +12,13 @@
 
 fn main() {
     let foo = &16;
+    //~^ HELP consider changing this to be a mutable reference
+    //~| SUGGESTION &mut 16
     *foo = 32;
+    //~^ ERROR cannot assign to `*foo` which is behind a `&` reference
     let bar = foo;
+    //~^ HELP consider changing this to be a mutable reference
+    //~| SUGGESTION &mut i32
     *bar = 64;
+    //~^ ERROR cannot assign to `*bar` which is behind a `&` reference
 }

--- a/src/test/ui/suggestions/issue-51515.rs
+++ b/src/test/ui/suggestions/issue-51515.rs
@@ -1,0 +1,18 @@
+// Copyright 2018 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+#![feature(nll)]
+
+fn main() {
+    let foo = &16;
+    *foo = 32;
+    let bar = foo;
+    *bar = 64;
+}

--- a/src/test/ui/suggestions/issue-51515.stderr
+++ b/src/test/ui/suggestions/issue-51515.stderr
@@ -1,14 +1,18 @@
 error[E0594]: cannot assign to `*foo` which is behind a `&` reference
-  --> $DIR/issue-51515.rs:15:5
+  --> $DIR/issue-51515.rs:17:5
    |
 LL |     let foo = &16;
    |               --- help: consider changing this to be a mutable reference: `&mut 16`
+...
 LL |     *foo = 32;
    |     ^^^^^^^^^ `foo` is a `&` reference, so the data it refers to cannot be written
 
 error[E0594]: cannot assign to `*bar` which is behind a `&` reference
-  --> $DIR/issue-51515.rs:17:5
+  --> $DIR/issue-51515.rs:22:5
    |
+LL |     let bar = foo;
+   |         --- help: consider changing this to be a mutable reference: `&mut i32`
+...
 LL |     *bar = 64;
    |     ^^^^^^^^^ `bar` is a `&` reference, so the data it refers to cannot be written
 

--- a/src/test/ui/suggestions/issue-51515.stderr
+++ b/src/test/ui/suggestions/issue-51515.stderr
@@ -1,0 +1,17 @@
+error[E0594]: cannot assign to `*foo` which is behind a `&` reference
+  --> $DIR/issue-51515.rs:15:5
+   |
+LL |     let foo = &16;
+   |               --- help: consider changing this to be a mutable reference: `&mut 16`
+LL |     *foo = 32;
+   |     ^^^^^^^^^ `foo` is a `&` reference, so the data it refers to cannot be written
+
+error[E0594]: cannot assign to `*bar` which is behind a `&` reference
+  --> $DIR/issue-51515.rs:17:5
+   |
+LL |     *bar = 64;
+   |     ^^^^^^^^^ `bar` is a `&` reference, so the data it refers to cannot be written
+
+error: aborting due to 2 previous errors
+
+For more information about this error, try `rustc --explain E0594`.

